Constrain dynamic api and impl

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96012
* #96338
* #96678

The goal of this PR is to build on top of https://github.com/pytorch/pytorch/pull/94787. Where that PR adds a mark_constrain, we add further refinement to it via mark_constrain_dynamic.

The interaction of mark_constrain_dynamic, config, eager/export, and more is complicated. For all intents and purposes, this API can be thought of as replacing mark_dynamic with a version of it that constrains the range of the given value.

There are a few logical flows here:

1) A user constraint appears on a dimension where we would normally accumulate no guards. In this case, we create a new guard matching the user specified range.

2) A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range. For example we check that dim `k > 5`. In this case, we (a) verify that the range does not contradict the guard (it does not) and we install both guards. In the future, here, would collapse all the ranges into a single guard. Something like https://github.com/pytorch/pytorch/pull/96144/files but not garbage.

3)  A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range, but that guard contradicts the user constraint. In this case, we raise.

4) In all other ways, this works exactly like mark_dynamic, and subsumes it, except for the major distinction in export. Export mark_dynamic does not allow for ANY constraining at all. With this API, export allows constraining with the bounds specified, same as (2) above.  


cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
[WIP][RFC] constrain dynamic api and impl

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/__init__.py:
```diff
@@ -157,13 +161,24 @@ def mark_dynamic(t, index):
-        t._dynamo_dynamic_indices.add(index)
+        new_range = MinMaxConstraint(min=min, max=max)
+        curr_range = t._dynamo_dynamic_indices[index]
+        t._dynamo_dynamic_indices[index] = MinMaxConstraint.INTERSECT(
+            curr_range, new_range
```
This actually needs to implement a disjoint check, since you can have disjoint ranges. It needs to:

1) check for overlap
2) If overlap, expand, if no overlap, new record

voznesenskym replied:
Actually, no lol

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

vadimkantorov commented:
Related?  https://github.com/pytorch/pytorch/issues/46948 https://github.com/pytorch/pytorch/issues/40373

voznesenskym commented:
> Related? #46948 #40373

Seems like it, at a distance. Also, Have you seen https://github.com/pytorch/pytorch/pull/94787?

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -194,7 +194,7 @@ def guard_scalar(a):
 
 # inclusive both ways
-def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
+def constrain_range(a, *, min: Optional[int], max: Optional[int] = None, user_directive=False):
```
Yes. I should just doc this whole thing.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1210,6 +1212,8 @@ def __init__(
         # range may contain ints which may not actually appear in
         # practice
         self.var_to_range: Dict["sympy.Symbol", ValueRanges] = {}
+        # var_to_range entries which were modified by manual user directive
```
Agreed.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1253,9 +1257,10 @@ def _produce_dyn_sizes(self, ex: torch.Tensor, source: Source) -> List[sympy.Exp
                     val, TensorPropertySource(source, TensorProperty.SIZE, i), is_dynamic
-                ))
+                )
+                size.append(new_sym)
```
This is the return `size` from `_produce_dyn_sizes` not the input `size` from `create_symbolic_sizes_strides_storage_offset` where it used to live.

voznesenskym commented on torch/_dynamo/__init__.py:
```diff
@@ -37,6 +41,7 @@
     "is_compiling",
     "register_backend",
     "list_backends",
+    "mark_dynamic_constrained",
```
Bikeshed with me, what names do you prefer?

voznesenskym commented:
Drafting this again as I would like to add a lot more context to the approach.

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

vadimkantorov commented:
> > Related? #46948 #40373
> 
> Seems like it, at a distance. Also, Have you seen #94787?

nope, but also that google doc is not public

in my original issues i proposed to somehow let the user express various tensor constraints and properties via assert statements that could then be taken into account by tracer

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1942,6 +2006,35 @@ def _should_allocate(user_marked_dynamic, assume_static_by_default):
 def _is_dim_dynamic(t, d):
     return hasattr(t, "_dynamo_dynamic_indices") and d in t._dynamo_dynamic_indices
 
+class MinMaxConstraint:
```
I don't really understand why we need another class for this

ezyang replied:
This seems overly specialized for min max constraint. Didn't we agree that there were more than one potential constraint type? Better just have a bucket constraint data type that has room to grow rather than call it MinMaxConstraint.

voznesenskym replied:
> Didn't we agree that there were more than one potential constraint type?

I thought we explicitly said we only have min max constrains?

ezyang replied:
min max to start, see https://docs.google.com/document/d/1aoIyYE8_6cYpWqS25thzVoIiKsT5aaUEOiiPwbIXt8k/edit?disco=AAAAsiJ5ujg

voznesenskym replied:
I thought we disagreed on that, and said we would punt it until we find a need. If this is indeed the case, I am happy to refactor.

gmagogsfm replied:
I think in the March 7th 11:45 PT discussion, we agreed on expressing min-max only and only add things when they are proven to be useful. 

In the referenced document, it says we may "add more stuff", what are included the "more stuff"?

ezyang commented on torch/_dynamo/__init__.py:
```diff
@@ -156,14 +162,96 @@ def mark_dynamic(t, index):
+
+
+@forbid_in_graph
+def mark_dynamic_constrain(
```
OK, with the docblock, I recommend just folding this into `mark_dynamic`, let's not have two APIs for this.

voznesenskym replied:
Sure, I am okay with that.

ezyang commented on torch/_dynamo/__init__.py:
```diff
@@ -156,14 +162,96 @@ def mark_dynamic(t, index):
+
+    We would raise.
+
+    This API behaves identically for eager and export.
```
When you're ready for nits, you need to properly format the code blocks in the docblock

voznesenskym replied:
I had them formatted, and doc ci failed, but it's still failing, so I need to do this differently.

ezyang commented on torch/_dynamo/__init__.py:
```diff
@@ -156,14 +162,96 @@ def mark_dynamic(t, index):
+
+    mark_dynamic_constrain allows users to provide a directive that the dimension will fall
+    within a given range. A range can be unbounded on either min, or max. Multiple calls to this API for
+    the same dimension will take the most conservative intersection of all ranges. At guard accumulation
```
Alternately, you could just error.

There is a little bit of an open UX question here, but erroring leaves the door open for either semantics.

ezyang replied:
Because the marker is EXTERNAL to the model, it is not obvious that intersection is better than overwriting. In particular, overwriting is more expressive than intersection.

voznesenskym replied:
No strong opinion! Can swap to overwriting.

voznesenskym replied:
discussed offline, swapped to first or clear.

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -501,13 +504,20 @@ def TENSOR_MATCH(self, guard: Guard):
                 if hasattr(value, "_dynamo_dynamic_indices"):
                     code.append(
-                        f"({tensor_name}._dynamo_dynamic_indices.issubset({value._dynamo_dynamic_indices})) if hasattr({tensor_name}, '_dynamo_dynamic_indices') else True"  # noqa: B950
+                        f"___dynamic_dims_check({tensor_name}, {dict(value._dynamo_dynamic_indices)})"
```
This is starting to look expensive. What if we do some sort of counter check (and when you update constraints, we bump the counter)?

voznesenskym replied:
Sorry, I don't understand this.

ezyang replied:
Yes. Update when we change metadata

voznesenskym replied:
Well, then you will never reuse past frames, right? That would be fine for invalidating the last frame, but it would be cache inefficient. 

I think we can move this to C++ though, since its per tensor, do you want me to follow up with that?

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -722,6 +732,7 @@ def compile_check_fn(
                 ("___check_tensors", check_tensors_fn),
                 ("___check_tensors_verbose", check_tensors_verbose_fn),
                 ("tensor_check_names", tensor_check_names),
+                ("oo", float("inf")),
```
Nope, figure out how to not do this. infinity should never be dumped in the guard code.

voznesenskym replied:
ack

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1210,6 +1233,14 @@ def __init__(
+        # for the purposes of protecting and enforcing user directives. In strict_mark_dyn mode, we do not
+        # allow ANY constraining of values for a given symbol. However, if this symbol is user constrained, we
+        # relax this requirement in favor of honoring the user directive and allowing the user specified range.
+        self.user_constrained: Set["sympy.Symbol"] = set()
```
I don't understand. All uses of `constrain_range` can only occur from "users". We do call constrain_range in nnz but this is to setup a "common sense" constrain range on unbacked symint. There are no uses of `constrain_range` otherwise.

voznesenskym replied:
We still need a structure to check, because every single symbol gets the "common sense" constraint. The presence of an external user invoked constrain changes the semantics of export + mark_dynamic.

Export + mark_dynamic --> no constraining allowed
Export + mark_constrain_dynamic --> constraining allowed within a range

Open to other ideas for doing this, but this was the simplest way WITHOUT checking if the range == "common sense" range. I figured just checking the "common sense" range was kinda grosser than this.

ezyang commented:
This seems like a good start, need to nail down some of the exact UX behavior.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1872,6 +1917,25 @@ def _simplify_floor_div(self, expr):
             self.evaluate_expr(eq_expr)
         return self.simplify(expr)
 
+    def _verify_valid_range(self, symbol, valid_range, expr):
```
As opposed to inline? Or opposed to what?

voznesenskym commented:
> This seems like a good start, need to nail down some of the exact UX behavior.

Thanks :). I agree the UX needs refinement. We have a discussion on the books for that tomorrow, let's see what folks decide. I do not hold any strong feelings about how it should be.

voznesenskym commented:
The new API we landed on is:

```
from torch._export import dynamic_dim

result = torch._dynamo.export(
    milan_dictation,
    *sixtyfour_tensors,
    constraints=[
        # if you do only dynamic_dim, this is sugar for
        # -Inf <= dynamic_dim(blah, 0) <= Inf; we don’t otherwise
        # permit direct int->bool conversion
        dynamic_dim(blah, 0),
        # operator overloading because it makes it clear whether
        # or not you’re inclusive-exclusive range or not
        0 <= dynamic_dim(blah, 1) <= 100,
    ]
)
```

The plan of action here is to implement it first for dynamic_dim, and then the API in this PR via operator overload

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"


The goal of this PR is to build on top of https://github.com/pytorch/pytorch/pull/94787. Where that PR adds a mark_constrain, we add further refinement to it via mark_constrain_dynamic.

The interaction of mark_constrain_dynamic, config, eager/export, and more is complicated. For all intents and purposes, this API can be thought of as replacing mark_dynamic with a version of it that constrains the range of the given value.

There are a few logical flows here:

1) A user constraint appears on a dimension where we would normally accumulate no guards. In this case, we create a new guard matching the user specified range.

2) A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range. For example we check that dim `k > 5`. In this case, we (a) verify that the range does not contradict the guard (it does not) and we install both guards. In the future, here, would collapse all the ranges into a single guard. Something like https://github.com/pytorch/pytorch/pull/96144/files but not garbage.

3)  A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range, but that guard contradicts the user constraint. In this case, we raise.

4) In all other ways, this works exactly like mark_dynamic, and subsumes it, except for the major distinction in export. Export mark_dynamic does not allow for ANY constraining at all. With this API, export allows constraining with the bounds specified, same as (2) above.  


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP][RFC] constrain dynamic api and impl"


The goal of this PR is to build on top of https://github.com/pytorch/pytorch/pull/94787. Where that PR adds a mark_constrain, we add further refinement to it via mark_constrain_dynamic.

The interaction of mark_constrain_dynamic, config, eager/export, and more is complicated. For all intents and purposes, this API can be thought of as replacing mark_dynamic with a version of it that constrains the range of the given value.

There are a few logical flows here:

1) A user constraint appears on a dimension where we would normally accumulate no guards. In this case, we create a new guard matching the user specified range.

2) A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range. For example we check that dim `k > 5`. In this case, we (a) verify that the range does not contradict the guard (it does not) and we install both guards. In the future, here, would collapse all the ranges into a single guard. Something like https://github.com/pytorch/pytorch/pull/96144/files but not garbage.

3)  A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range, but that guard contradicts the user constraint. In this case, we raise.

4) In all other ways, this works exactly like mark_dynamic, and subsumes it, except for the major distinction in export. Export mark_dynamic does not allow for ANY constraining at all. With this API, export allows constraining with the bounds specified, same as (2) above.  


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Constrain dynamic api and impl"


The goal of this PR is to build on top of https://github.com/pytorch/pytorch/pull/94787. Where that PR adds a mark_constrain, we add further refinement to it via mark_constrain_dynamic.

The interaction of mark_constrain_dynamic, config, eager/export, and more is complicated. For all intents and purposes, this API can be thought of as replacing mark_dynamic with a version of it that constrains the range of the given value.

There are a few logical flows here:

1) A user constraint appears on a dimension where we would normally accumulate no guards. In this case, we create a new guard matching the user specified range.

2) A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range. For example we check that dim `k > 5`. In this case, we (a) verify that the range does not contradict the guard (it does not) and we install both guards. In the future, here, would collapse all the ranges into a single guard. Something like https://github.com/pytorch/pytorch/pull/96144/files but not garbage.

3)  A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range, but that guard contradicts the user constraint. In this case, we raise.

4) In all other ways, this works exactly like mark_dynamic, and subsumes it, except for the major distinction in export. Export mark_dynamic does not allow for ANY constraining at all. With this API, export allows constraining with the bounds specified, same as (2) above.  


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Constrain dynamic api and impl"


The goal of this PR is to build on top of https://github.com/pytorch/pytorch/pull/94787. Where that PR adds a mark_constrain, we add further refinement to it via mark_constrain_dynamic.

The interaction of mark_constrain_dynamic, config, eager/export, and more is complicated. For all intents and purposes, this API can be thought of as replacing mark_dynamic with a version of it that constrains the range of the given value.

There are a few logical flows here:

1) A user constraint appears on a dimension where we would normally accumulate no guards. In this case, we create a new guard matching the user specified range.

2) A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range. For example we check that dim `k > 5`. In this case, we (a) verify that the range does not contradict the guard (it does not) and we install both guards. In the future, here, would collapse all the ranges into a single guard. Something like https://github.com/pytorch/pytorch/pull/96144/files but not garbage.

3)  A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range, but that guard contradicts the user constraint. In this case, we raise.

4) In all other ways, this works exactly like mark_dynamic, and subsumes it, except for the major distinction in export. Export mark_dynamic does not allow for ANY constraining at all. With this API, export allows constraining with the bounds specified, same as (2) above.  


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
I think I have a few bugs in the mechanism :( back to the drawing board

voznesenskym committed:
Update on "Constrain dynamic api and impl"


The goal of this PR is to build on top of https://github.com/pytorch/pytorch/pull/94787. Where that PR adds a mark_constrain, we add further refinement to it via mark_constrain_dynamic.

The interaction of mark_constrain_dynamic, config, eager/export, and more is complicated. For all intents and purposes, this API can be thought of as replacing mark_dynamic with a version of it that constrains the range of the given value.

There are a few logical flows here:

1) A user constraint appears on a dimension where we would normally accumulate no guards. In this case, we create a new guard matching the user specified range.

2) A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range. For example we check that dim `k > 5`. In this case, we (a) verify that the range does not contradict the guard (it does not) and we install both guards. In the future, here, would collapse all the ranges into a single guard. Something like https://github.com/pytorch/pytorch/pull/96144/files but not garbage.

3)  A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range, but that guard contradicts the user constraint. In this case, we raise.

4) In all other ways, this works exactly like mark_dynamic, and subsumes it, except for the major distinction in export. Export mark_dynamic does not allow for ANY constraining at all. With this API, export allows constraining with the bounds specified, same as (2) above.  


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Constrain dynamic api and impl"


The goal of this PR is to build on top of https://github.com/pytorch/pytorch/pull/94787. Where that PR adds a mark_constrain, we add further refinement to it via mark_constrain_dynamic.

The interaction of mark_constrain_dynamic, config, eager/export, and more is complicated. For all intents and purposes, this API can be thought of as replacing mark_dynamic with a version of it that constrains the range of the given value.

There are a few logical flows here:

1) A user constraint appears on a dimension where we would normally accumulate no guards. In this case, we create a new guard matching the user specified range.

2) A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range. For example we check that dim `k > 5`. In this case, we (a) verify that the range does not contradict the guard (it does not) and we install both guards. In the future, here, would collapse all the ranges into a single guard. Something like https://github.com/pytorch/pytorch/pull/96144/files but not garbage.

3)  A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range, but that guard contradicts the user constraint. In this case, we raise.

4) In all other ways, this works exactly like mark_dynamic, and subsumes it, except for the major distinction in export. Export mark_dynamic does not allow for ANY constraining at all. With this API, export allows constraining with the bounds specified, same as (2) above.  


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Constrain dynamic api and impl"


The goal of this PR is to build on top of https://github.com/pytorch/pytorch/pull/94787. Where that PR adds a mark_constrain, we add further refinement to it via mark_constrain_dynamic.

The interaction of mark_constrain_dynamic, config, eager/export, and more is complicated. For all intents and purposes, this API can be thought of as replacing mark_dynamic with a version of it that constrains the range of the given value.

There are a few logical flows here:

1) A user constraint appears on a dimension where we would normally accumulate no guards. In this case, we create a new guard matching the user specified range.

2) A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range. For example we check that dim `k > 5`. In this case, we (a) verify that the range does not contradict the guard (it does not) and we install both guards. In the future, here, would collapse all the ranges into a single guard. Something like https://github.com/pytorch/pytorch/pull/96144/files but not garbage.

3)  A user constraint appears on a dimension where we would accumulate a guard fully encompassed by a range, but that guard contradicts the user constraint. In this case, we raise.

4) In all other ways, this works exactly like mark_dynamic, and subsumes it, except for the major distinction in export. Export mark_dynamic does not allow for ANY constraining at all. With this API, export allows constraining with the bounds specified, same as (2) above.  


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
https://github.com/pytorch/pytorch/pull/96699
